pydis_site/apps/api/viewsets/bot/nomination.py

```diff
@@ -1,32 +1,224 @@
+
+    See operation 1 for the response format and status codes.
+    """
     serializer_class = NominationSerializer
-    queryset = Nomination.objects.prefetch_related('author', 'user')
-    frozen_fields = ('author', 'inserted_at', 'user')
+    queryset = Nomination.objects.all()
+    filter_backends = (DjangoFilterBackend, SearchFilter, OrderingFilter)
+    filter_fields = ('user__id', 'actor__id', 'active')
+    frozen_fields = ('id', 'actor', 'inserted_at', 'user', 'ended_at')
+    frozen_on_create = ('ended_at', 'end_reason', 'active', 'inserted_at')
 
-    def update(self, request, *args, **kwargs):
+    def create(self, request, *args, **kwargs):
         """
-        DRF method for updating a Nomination.
+        DRF method for creating a Nomination.
 
         Called by the Django Rest Framework in response to the corresponding HTTP request.
         """
+        for field in request.data:
+            if field in self.frozen_on_create:
+                raise ValidationError({field: ['This field cannot be set at creation.']})
 
+        user_id = request.data.get("user")
+        if Nomination.objects.filter(active=True, user__id=user_id).exists():
+            raise ValidationError({'active': ['There can only be one active nomination.']})
+
+        serializer = self.get_serializer(
+            data=ChainMap(
+                request.data,
+                {"active": True}
+            )
+        )
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+
+    def partial_update(self, request, *args, **kwargs):
+        """
+        DRF method for updating a Nomination.
+
+        Called by the Django Rest Framework in response to the corresponding HTTP request.
+        """
         for field in request.data:
             if field in self.frozen_fields:
                 raise ValidationError({field: ['This field cannot be updated.']})
 
         instance = self.get_object()
         serializer = self.get_serializer(instance, data=request.data, partial=True)
         serializer.is_valid(raise_exception=True)
+
+        data = serializer.validated_data
+
+        # There are three distinct PATCH scenarios we need to validate.
+        if instance.active and 'active' not in data:
+            # 1. We're updating an active nomination without ending it.
+            if request.data.get('end_reason'):
+                raise ValidationError(
+                    {'end_reason': ["An active nomination can't have an end reason."]}
+                )
+
+        elif instance.active and not data.get('active'):
+            # 2. We're ending an active nomination.
+            if 'reason' in data:
+                raise ValidationError(
+                    {'reason': ['This field cannot be set when ending a nomination.']}
```

Comment:
Why not? To prevent accidentally setting the wrong reason?

Reply:
Because ending a nomination is a semantically different action than updating the original reason for the nomination. I feel that we should separate these "update the current information" from "changing the status of a nomination" operations.

Reply:
I disagree with that viewpoint. It's been merged into one endpoint yet you still treat it as if its two separate ones. The way I see it, a PATCH is a PATCH - simply an update of the model. We should allow updating anything one wants so long as it doesn't break the logic of the model.

If hypothetically one wishes to edit `reason` and end a nomination simultaneously, you are suggesting they make two separate requests to the same endpoint? Forgive me, but that just seems silly to me.

Reply:
If having a single endpoint means a single type of action, then merging the two endpoints was a mistake, in my opinion. They are two very different actions with very different consequences in terms of what happens in the database.

When you update information, you're only updating the information that is currently there. When you end a nomination, you're not just updating information, but setting additional fields that for all intents and purposes were considered unset up until that point. Ending a nomination evens sets a field that's not provided by the request, namely the end date of the nomination.

When we just consider it to be all "updating fields", what actually happens in the database can't be predicted by just the request. The same request may or may not modify the `end_at` field, depending on the current and unknown status of the nomination. That doesn't feel right to me.

Reply:
Then would you rather revert commits to have the endpoints separate? I was on the fence about that begin with, it was more @jchristgit's suggestion.

Reply:
I don't really care what direction we choose, I just don't like two identical requests triggering different behavior in the back-end without the user having explicit control over it or being able to predict what's going to happen.